src/libclipper/test/json_util_test.cpp

```diff
@@ -0,0 +1,206 @@
+}
+
+TEST(JsonUtilTests, TestArrayTypeMismatch) {
+  // JSON arrays may contain items of different types so the array
+  //  parsing methods should throw an exception if this is the case.
+  std::string source_json =
+      "{\"double_array_with_int\": [1.2, 3.4, 5, 6.7],"
+       "\"int_array_with_string\": [1, 2, 3, \"4\"],"
+       "\"string_array_with_bool\": [\"test1\", \"phrase2\", true]}";
+
+  rapidjson::Document d;
+  parse_json(source_json, d);
+  ASSERT_THROW(
+        get_double_array(d, "double_array_with_int"),
+        json_semantic_error);
+  ASSERT_THROW(
+        get_int_array(d, "int_array_with_string"),
+        json_semantic_error);
+  ASSERT_THROW(
+        get_string_array(d, "string_array_with_bool"),
+        json_semantic_error);
+}
+
+TEST(JsonUtilTests, TestEmptyDocumentToString) {
+  rapidjson::Document d;
+  EXPECT_EQ("null", to_json_string(d));
+  d.SetObject();
+  EXPECT_EQ("{}", to_json_string(d));
+  d.SetArray();
+  EXPECT_EQ("[]", to_json_string(d));
+  d.SetDouble(0.3);
+  EXPECT_EQ("0.3", to_json_string(d));
+  d.SetBool(true);
+  EXPECT_EQ("true", to_json_string(d));
+}
+
+TEST(JsonUtilTests, TestAddToNonObject) {
+  rapidjson::Document d;
+  d.SetArray();
+  ASSERT_THROW(add_int(3, "int_val", d), json_semantic_error);
+}
+
+TEST(JsonUtilTests, TestParseCandidateModels) {
+  // Semantic checking to assert that model_version > 0?
+  std::string correct_json =
+      "{\"correct_candidate_models\": ["
+          "{\"model_name\": \"sklearn_svm\", \"model_version\": 1},"
+          "{\"model_name\": \"sklearn_svm\", \"model_version\": 2},"
+          "{\"model_name\": \"network\", \"model_version\": 3}"
+       "]}";
+  std::string missing_name_json =
+      "{\"missing_name\": [{\"model_version\": 3}]}";
+  std::string missing_version_json =
+      "{\"missing_version\": [{\"model_name\": \"m\"}]}";
+  std::string wrong_name_type_json =
+      "{\"wrong_name_type\": [{\"model_name\": 123, \"model_version\": 1}]}";
+  std::string wrong_version_type_json =
+      "{\"wrong_version_type\": ["
+          "{\"model_name\": \"g\", \"model_version\": \"123\"}"
+       "]}";
+  std::vector<clipper::VersionedModelId> expected_models =
+      {{"sklearn_svm", 1}, {"sklearn_svm", 2}, {"network", 3}};
+
+  rapidjson::Document d;
+  parse_json(correct_json, d);
+  EXPECT_EQ(
+        get_candidate_models(d, "correct_candidate_models"),
+        expected_models);
+
+  parse_json(missing_name_json, d);
+  ASSERT_THROW(get_candidate_models(d, "missing_name"), json_semantic_error);
+
+  parse_json(missing_version_json, d);
+  ASSERT_THROW(get_candidate_models(d, "missing_version"), json_semantic_error);
+
+  parse_json(wrong_name_type_json, d);
+  ASSERT_THROW(get_candidate_models(d, "wrong_name_type"), json_semantic_error);
+
+  parse_json(wrong_version_type_json, d);
+  ASSERT_THROW(
+        get_candidate_models(d, "wrong_version_type"),
+        json_semantic_error);
+}
+
+TEST(JsonUtilTests, TestParseNestedObject) {
```

Comment:
@Corey-Zumar Let me know if you have any feedback on this example JSON deserialization with nested JSON objects.